tests/test_scheduler.py

```diff
@@ -0,0 +1,136 @@
+            mdl.experimentQueue.extend(data)
+            self.assertEqual(mdl.rowCount(), len(data))
+            for i, exp in enumerate(data):
+                self.assertEqual(mdl.data(mdl.index(i)), exp)
+
+    def test_drag_and_drop(self):
+        mdl = scheduler.ExperimentModel()
+        data = [ExperimentInfo("1", {"rid": 1, "priority": 2}),
+                ExperimentInfo("2", {"rid": 2, "priority": 1}),
+                ExperimentInfo("3", {"rid": 3, "priority": 1})
+               ]
+        mdl.experimentQueue.extend(data)
+        mime0 = QMimeData()
+        mime0.setText("0")
+        mime1 = QMimeData()
+        mime1.setText("1")
+        mime2 = QMimeData()
+        mime2.setText("2")
+        mdl.dropMimeData(mime0, Qt.MoveAction, 0, 0, mdl.index(0)) # exp1 above exp1
+        self.assertEqual(mdl.experimentQueue, data)
+        mdl.dropMimeData(mime0, Qt.MoveAction, 2, 0, mdl.index(0)) # exp1 above exp3
+        self.assertEqual(mdl.experimentQueue, data)
+        mdl.dropMimeData(mime1, Qt.MoveAction, 3, 0, mdl.index(0)) # exp2 below exp3
+        self.assertEqual(mdl.experimentQueue, [data[0], data[2], data[1]])
+        mdl.dropMimeData(mime2, Qt.MoveAction, 1, 0, mdl.index(0)) # exp2 above exp3
+        self.assertEqual(mdl.experimentQueue, data)
+
+
+class SchedulerAppTest(unittest.TestCase):
+    """Unit tests for SchedulerApp class."""
+
+    def setUp(self):
+        self.qapp = QApplication([])
+
+    def tearDown(self):
+        del self.qapp
+
+    def test_add_experiment(self):
+        app = scheduler.SchedulerApp(name="name")
+        with mock.patch.object(app.schedulerFrame.model, "experimentQueue") as mockedQueue:
+            data = [ExperimentInfo(str(i), {"rid": i, "priority": 10 - i}) for i in range(10)]
+            for info in data:
+                app.addExperiment(info)
+                mockedQueue.append.assert_called_with(info)
+
+    def test_run_experiment(self):
+        app = scheduler.SchedulerApp(name="name")
+        with mock.patch.object(app.schedulerFrame, "runningView") as mockedView:
+            info = ExperimentInfo("1", {"rid": 1, "priority": 1})
+            app.runExperiment(info)
+            mockedView.updateInfo.assert_called_with(info)
+
+
+class SchedulerFunctionalTest(unittest.TestCase):
+    """Functional tests for SchedulerApp class."""
+
+    def setUp(self):
+        self.qapp = QApplication([])
+
+    def tearDown(self):
+        del self.qapp
+
+    def test_add_experiment(self):
+        app = scheduler.SchedulerApp(name="name", parent=QObject())
+        data = [ExperimentInfo(str(i), {"rid": i, "priority": 10 - i}) for i in range(10)]
+        for info in data:
+            app.addExperiment(info)
+        self.assertEqual(app.schedulerFrame.model.experimentQueue, data)
+        app.schedulerFrame.model.experimentQueue = []
+        permutation = [1, 9, 3, 8, 7, 4, 2, 6, 5, 0]
+        inv_permutation = [1, 3, 4, 7, 8, 5, 2, 6, 0, 9]
```

Comment:
Could you explain more about `permutation` and `inv_permutation`?

Reply:
`permutation` is the priority mix-up, and the `inv_permutation` is the answer for this mix-up.

I will mention this in the code.

Reply:
Thank you, now I understood.
How about using more intuitive names?
I understand what you intended, but it is unnecessarily mathematic (which requires more backgrounds for code readers).
For example, `priorities` and `sorted_indices`.